### PR TITLE
Copr support streaming in coprocessor Batch Mode

### PR DESCRIPTION
**Pull request by mapleFU**
###  What have you changed?

Support streaming in TiKV Batch Mode.

#### Summarize your change

Support streaming in TiKV coprocessor. Here was 3 part of the change:
1. Support streaming in `BatchDAGBuilder`, and open the support for it in `build_handler`.
2. Implements ` handle_streaming_request` in `components/tidb_query/src/batch/runner.rs`
3. Implements `take_scanned_range` for `BatchExecutorRunner ` and add correspond integration tests for it (#5362 ).
4. Update testing in `test_stream_batch_row_limit`.

#### How does this PR work? Include a brief introduction for any changed logic (optional)

Please "1" in last part.

###  What is the type of the changes?

Pick one of the following and delete the others:

- New feature (Support streaming in coprocessor Batch Mode)

###  How is the PR tested?

Updated integration test.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

I don't know now.

###  Does this PR affect `tidb-ansible`?

No.

###  Refer to a related PR or issue link (optional)

* Pull request for `take_scanned_range` in #5362 .

###  Benchmark result if necessary (optional)

Haven't done yet.

### CODE REVIEW

**Comment by mapleFU** [comment]
`tikv=master` still failed... OMG

**Comment by mapleFU** [comment]
/run-integration-tests tidb=pr/11949
--

**Comment by mapleFU** [comment]
Seems that problem is `TiKV Server timeout`, but I don't know why the error happens.

File: components/tidb_query/src/batch/runner.rs [hunk 1]
```diff
@@ -452,4 +461,140 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        let (mut record_cnt, mut is_drained) = (0, false);
+        let mut already_read_cnt = 0;
+        let mut chunk = Chunk::default();
+
+        // record count less than batch size and is not drained
+        info!(
+            "handle_streaming_request in Batch called, self.stream_batch_row_limit is {}",
+            self.stream_batch_row_limit
+        );
+        while record_cnt < 32 && already_read_cnt < 32 && !is_drained {
```
Comment (lonng) [inline components/tidb_query/src/batch/runner.rs]: Prefer `self.out_most_executor.next_batch(stream_batch_row_limit)` directly instead of this `while` loop.

File: src/coprocessor/dag/mod.rs [hunk 2]
```diff
@@ -23,22 +23,34 @@ pub fn build_handler<S: Store + 'static>(
...
             is_batch = true;
         }
     }
 
     if is_batch {
         COPR_DAG_REQ_COUNT.with_label_values(&["batch"]).inc();
-        Ok(BatchDAGHandler::new(req, ranges, store, deadline)?.into_boxed())
+        info!("Build batch executor with streaming {}", is_streaming);
+        Ok(
+            BatchDAGHandler::new(req, ranges, store, deadline, batch_row_limit, is_streaming)?
```
Comment (lonng) [inline src/coprocessor/dag/mod.rs]: I think we don't need the `is_streaming` parameter anymore.

**Comment by mapleFU** [comment]
Seems that `self.is_scanned_range_aware` is not configed.

**Comment by mapleFU** [comment]
Unit test `test_stream_batch_row_limit` set  `let stream_row_limit = 2;` and test it strictly.

So here maybe I need to move this test? Or change the way I implement the batch streaming?

File: components/tidb_query/src/batch/runner.rs (hunk 1 above)
Comment (mapleFU) [reply components/tidb_query/src/batch/runner.rs]: > Prefer `self.out_most_executor.next_batch(stream_batch_row_limit)` directly instead of this `while` loop.

1. Here we want the final result both no less than `stream_batch_row_limit` and not be a lot more than the real result.
2. `self.out_most_executor.next_batch(stream_batch_row_limit)` will fetch `stream_batch_row_limit` `physical_columns` , but the logical columns are tend to be less than `stream_batch_row_limit`.

File: components/tidb_query/src/batch/executors/index_scan_executor.rs [hunk 3]
```diff
@@ -28,13 +28,14 @@ impl BatchIndexScanExecutor<Box<dyn Storage<Statistics = ()>>> {
 }
 
 impl<S: Storage> BatchIndexScanExecutor<S> {
-    pub fn new(
+    pub fn new_with_scanned_range_aware(
```
Comment (breeswish) [inline components/tidb_query/src/batch/executors/index_scan_executor.rs]: I think adding a parameter is enough. No need to introduce another function.

File: components/tidb_query/src/batch/executors/table_scan_executor.rs [hunk 4]
```diff
@@ -32,12 +32,13 @@ impl BatchTableScanExecutor<Box<dyn Storage<Statistics = ()>>> {
 }
 
 impl<S: Storage> BatchTableScanExecutor<S> {
-    pub fn new(
+    pub fn new_with_scanned_range_aware(
```
Comment (breeswish) [inline components/tidb_query/src/batch/executors/table_scan_executor.rs]: ditto.

File: components/tidb_query/src/batch/runner.rs [hunk 5]
```diff
@@ -109,6 +115,8 @@ pub fn build_executors<S: Storage + 'static, C: ExecSummaryCollector + 'static>(
     storage: S,
     ranges: Vec<KeyRange>,
     config: Arc<EvalConfig>,
+    // TODO: apply this on the system
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: TODO?

File: components/tidb_query/src/batch/runner.rs [hunk 6]
```diff
@@ -46,6 +47,11 @@ pub struct BatchExecutorsRunner<SS> {
     collect_exec_summary: bool,
 
     exec_stats: ExecuteStats,
+
+    stream_batch_row_limit: usize,
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: prefer to give comments to these two variables describing their roles, considering that these two variables' names are very similar.

Comment (breeswish) [reply components/tidb_query/src/batch/runner.rs]: A rename for stream_batch_row_limit globally may also be considered. e.g. stream_min_rows_each_iter

File: components/tidb_query/src/batch/runner.rs [hunk 7]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
-            }
+            grow_batch_size(&mut batch_size);
         }
     }
 
     pub fn collect_storage_stats(&mut self, dest: &mut SS) {
         self.out_most_executor.collect_storage_stats(dest);
     }
+
+    // TODO: do not copy from the origin
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: TODO?

File: components/tidb_query/src/batch/runner.rs [hunk 8]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        if !is_drained || record_cnt > 0 {
+            let range = self.out_most_executor.take_scanned_range();
+            return self
+                .make_stream_response(chunk, warnings)
+                .map(|r| (Some((r, range)), is_drained));
+        }
+        Ok((None, true))
+    }
+
+    // TODO: check if this method can be share or put in utils
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: TODO?

File: components/tidb_query/src/batch/runner.rs [hunk 9]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        &mut self,
+        chunk: Chunk,
+        warnings: EvalWarnings,
+    ) -> Result<StreamResponse> {
+        self.out_most_executor
+            .collect_exec_stats(&mut self.exec_stats);
+
+        let mut s_resp = StreamResponse::default();
+        s_resp.set_data(box_try!(chunk.write_to_bytes()));
+        // TODO: now how to fill this and give warnings to Response
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: TODO?

File: components/tidb_query/src/batch/runner.rs [hunk 10]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+
+        Ok(s_resp)
+    }
+
+    /// return (is_drained, EvalWarnings, record_cnt)
+    fn internal_handle_request(
+        &mut self,
+        batch_size: usize,
+        chunk: &mut Chunk,
+    ) -> Result<(bool, EvalWarnings, usize)> {
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: Too many return types.

File: components/tidb_query/src/batch/runner.rs [hunk 11]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
     }
+
+    // TODO: do not copy from the origin
+    pub fn handle_streaming_request(
+        &mut self,
+    ) -> Result<(Option<(StreamResponse, IntervalRange)>, bool)> {
+        let mut warnings = self.config.new_eval_warnings();
+
+        // if the data read finished, is_drained means "finished"
+        let (mut record_cnt, mut is_drained) = (0, false);
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: For the new code, please use `record_len` or `record_count` instead of `record_cnt`.

File: components/tidb_query/src/batch/runner.rs [hunk 12]
```diff
@@ -440,16 +426,128 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        let mut chunk = Chunk::default();
+
+        // record count less than batch size and is not drained
+        while record_cnt < self.stream_batch_row_limit && !is_drained {
+            let (drained, mut result_warnings, cnt) =
+                self.internal_handle_request(self.stream_batch_size, &mut chunk)?;
+            record_cnt += cnt;
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: ditto.

File: src/coprocessor/dag/mod.rs [hunk 13]
```diff
@@ -75,6 +78,7 @@ impl RequestHandler for DAGHandler {
     }
 
     fn handle_streaming_request(&mut self) -> Result<(Option<Response>, bool)> {
+        info!("None batch DAGHandler stream called.");
```
Comment (breeswish) [inline src/coprocessor/dag/mod.rs]: ?

File: tests/integrations/coprocessor/test_select.rs [hunk 14]
```diff
@@ -105,46 +101,19 @@ fn test_stream_batch_row_limit() {
...
-            let result_encoded = datum::encode_value(&row).unwrap();
-            assert_eq!(result_encoded, &*expected_encoded);
+        if i == length - 1 {
+            break;
         }
+        let cnt = resp.get_output_counts();
+        assert!(
+            [stream_row_limit as i64].as_ref() <= cnt
+                && [(stream_row_limit + BATCH_MAX_SIZE) as i64].as_ref() >= cnt
+        );
```
Comment (breeswish) [inline tests/integrations/coprocessor/test_select.rs]: You should also check contents, not only just row counts.

File: components/tidb_query/src/batch/runner.rs [hunk 15]
```diff
@@ -374,14 +374,12 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
-                self.internal_handle_request(batch_size, &mut chunk)?;
+            let (is_drained, record_len) =
+                // return (is_drained, record_len)
+                self.internal_handle_request(batch_size, &mut chunk, &mut warnings)?;
 
-            warnings.merge(&mut result_warnings);
-
-            if record_cnt > 0 {
+            if record_len > 0 {
                 chunks.push(chunk);
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: so why not make `internal_handle_request` push chunk directly?

**Comment by mapleFU** [comment]
/run-integration-tests tidb=pr/11949
--

File: components/tidb_query/src/batch/runner.rs (hunk 15 above)
Comment (mapleFU) [reply components/tidb_query/src/batch/runner.rs]: > so why not make `internal_handle_request` push chunk directly?

In streaming mode, only one `chunk` variable is created, and `internal_handle_request` will push data to it. And in non-streaming mode, a `Vec<Chunk>` is used. So only in non-streaming mode, we need to use `chunks.push(chunk);`.

File: components/test_coprocessor/src/fixture.rs [hunk 16]
```diff
@@ -56,23 +56,27 @@ pub fn init_data_with_engine_and_commit<E: Engine>(
     init_data_with_details(ctx, engine, tbl, vals, commit, &Config::default())
 }
 
-pub fn init_data_with_details<E: Engine>(
+pub fn init_data_with_details<E: Engine, T: AsRef<str>>(
```
Comment (breeswish) [inline components/test_coprocessor/src/fixture.rs]: why is this file changed?

File: components/tidb_query/src/batch/runner.rs [hunk 17]
```diff
@@ -46,6 +47,12 @@ pub struct BatchExecutorsRunner<SS> {
     collect_exec_summary: bool,
 
     exec_stats: ExecuteStats,
+
+    /// Minimum rows to return in batch stream mode.
+    stream_min_rows_each_iter: usize,
+
+    /// `batch_size` in strReam mode, this variable will be initialized as `BATCH_INITIAL_SIZE`.
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: typo: strReam

Comment (breeswish) [reply components/tidb_query/src/batch/runner.rs]: At the time reading this comment, there is no `batch_size`. So you still need to explain what is `batch_size`, or at least give some reference.

File: components/tidb_query/src/batch/runner.rs [hunk 18]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
             }
 
             // Grow batch size
-            if batch_size < BATCH_MAX_SIZE {
-                batch_size *= BATCH_GROW_FACTOR;
-                if batch_size > BATCH_MAX_SIZE {
-                    batch_size = BATCH_MAX_SIZE
-                }
-            }
+            grow_batch_size(&mut batch_size);
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: looks like you can grow batch size in `internal_handle_request` so that there is no need for a standalone function `grow_batch_size`.

File: components/tidb_query/src/batch/runner.rs [hunk 19]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
     pub fn collect_storage_stats(&mut self, dest: &mut SS) {
         self.out_most_executor.collect_storage_stats(dest);
     }
+
+    pub fn handle_streaming_request(
+        &mut self,
+    ) -> Result<(Option<(StreamResponse, IntervalRange)>, bool)> {
+        let mut warnings = self.config.new_eval_warnings();
+
+        // if the data read finished, is_drained means "finished"
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: I think is_drained means finished even if data read is not finished?

File: components/tidb_query/src/batch/runner.rs [hunk 20]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        let (mut record_len, mut is_drained) = (0, false);
+        let mut chunk = Chunk::default();
+
+        // record count less than batch size and is not drained
+        while record_len < self.stream_min_rows_each_iter && !is_drained {
+            let (drained, cnt) =
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: change `cnt` to `len` or `count`.

File: components/tidb_query/src/batch/runner.rs [hunk 21]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+                self.internal_handle_request(self.stream_batch_size, &mut chunk, &mut warnings)?;
+            record_len += cnt;
+            is_drained = drained;
+
+            // Grow batch size
+            grow_batch_size(&mut self.stream_batch_size)
+        }
+
+        if !is_drained || record_len > 0 {
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: what if `is_drained && record_len > 0`?

File: components/tidb_query/src/batch/runner.rs [hunk 22]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+
+        s_resp.set_warnings(warnings.warnings.into());
+        s_resp.set_warning_count(warnings.warning_cnt as i64);
+
+        self.exec_stats.clear();
+
+        Ok(s_resp)
+    }
+
+    /// return (is_drained, EvalWarnings, record_cnt)
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: Outdated comment. also please use a more formal comment format. See https://rust-lang-nursery.github.io/api-guidelines/documentation.html

File: components/tidb_query/src/batch/runner.rs [hunk 23]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        warnings: &mut EvalWarnings,
+    ) -> Result<(bool, usize)> {
+        let is_drained;
+        let mut record_cnt = 0;
+
+        self.deadline.check()?;
+
+        let mut result = self.out_most_executor.next_batch(batch_size);
+
+        // fill is_drained
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: Incorrect comment: but you are not filling `is_drained` when `result.is_drained` is `Err(_)`

File: components/tidb_query/src/batch/runner.rs [hunk 24]
```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+
+    /// return (is_drained, EvalWarnings, record_cnt)
+    fn internal_handle_request(
+        &mut self,
+        batch_size: usize,
+        chunk: &mut Chunk,
+        warnings: &mut EvalWarnings,
+    ) -> Result<(bool, usize)> {
+        let is_drained;
+        let mut record_cnt = 0;
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: Either `record_count` or `record_len`.

File: src/coprocessor/dag/mod.rs [hunk 25]
```diff
@@ -91,13 +94,17 @@ impl BatchDAGHandler {
         ranges: Vec<KeyRange>,
         store: S,
         deadline: Deadline,
+        stream_batch_row_limit: usize,
```
Comment (breeswish) [inline src/coprocessor/dag/mod.rs]: missing rename?

File: components/tidb_query/src/batch/runner.rs (hunk 21 above)
Comment (mapleFU) [reply components/tidb_query/src/batch/runner.rs]: > is_drained && record_len > 0

what if `is_drained && record_len == 0`,  it will return `Ok((None, true))`. 

If `is_drained && record_len > 0`, it returns the ` ((data, range), true)` because it's drained.

**Comment by mapleFU** [comment]
Seems that sometimes it will read more data than we expected, I will add some log for testing.

**Comment by mapleFU** [comment]
Sometimes `Integration MySQL Test` and `Integration MySQL Test Cached` in `idc-jenkins-ci-tikv/integration-common-test` would failed:

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration_common_test/detail/tikv_ghpr_integration_common_test/2465/pipeline/150

**Comment by mapleFU** [comment]
The SQL and DDL in ` Integration MySQL Test` which would raise error are:

``` sql
CREATE TABLE t2 (a CHAR(2), b CHAR(2), c CHAR(2), INDEX (a, b));
INSERT INTO t2 (a, b) SELECT i, i FROM t1;
INSERT INTO t2 (a, b) SELECT t1.i, t1.i FROM t1, t1 x1, t1 x2;

#FLUSH STATUS;
SELECT * FROM t2 ORDER BY a, b LIMIT 5;
# SHOW SESSION STATUS LIKE 'Sort%';
# SHOW STATUS LIKE 'Handler_read_%';

# FLUSH STATUS;
SELECT * FROM t2 ORDER BY a DESC, b DESC LIMIT 5;
# SHOW SESSION STATUS LIKE 'Sort%';
# SHOW STATUS LIKE 'Handler_read_%';

# FLUSH STATUS;
DELETE FROM t2 ORDER BY a DESC, b DESC LIMIT 5;
# SHOW SESSION STATUS LIKE 'Sort%';
# SHOW STATUS LIKE 'Handler_read_%';
SELECT * FROM t2 WHERE c = 10 ORDER BY a DESC, b DESC;

DROP TABLE t1, t2;
```

Seems that we read from the snapshot before we delete the rows...

**Comment by mapleFU** [comment]
Using this test.t: https://gist.github.com/mapleFU/4db459aeae4f6fd59d35d35777022cb3
This result.r: https://gist.github.com/mapleFU/ba4cf81c25639d7d6ed8a71f59fedfc9

Can raise the errors for us.

**Comment by breeswish** [comment]
Are you still resolving the bug?

**Comment by mapleFU** [comment]
> Are you still resolving the bug?

If using tidb pr11949, and change tikv branch to master and use coprocessor non-batch streaming, we may face the same problem. I think thats a tidb error, but I don't know how to fix it.

**Comment by breeswish** [comment]
@mapleFU So @SunRunAway should fix it before we continue.

**Comment by mapleFU** [comment]
I create an issue in TiDB, and waiting for them to fix it: https://github.com/pingcap/tidb/issues/12131

File: components/tidb_query/src/batch/runner.rs [hunk 26]
```diff
@@ -440,16 +425,122 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
...
+        warnings: &mut EvalWarnings,
+    ) -> Result<(bool, usize)> {
+        let is_drained;
+        let mut record_len = 0;
+
+        self.deadline.check()?;
+
+        let mut result = self.out_most_executor.next_batch(batch_size);
+
+        match result.is_drained {
```
Comment (lonng) [inline components/tidb_query/src/batch/runner.rs]: Prefer: 
```rust
let is_drained = result.is_drained?;
```

Comment (mapleFU) [reply components/tidb_query/src/batch/runner.rs]: > Prefer:
> 
> ```rust
> let is_drained = result.is_drained?;
> ```

Yes, I'll change it.

**Comment by breeswish** [comment]
How is the integration test? Could you discover what TiDB PR fixes the issue?

**Comment by lonng** [comment]
@breeswish This BUG has not been fixed yet in TiDB. I am investigating.

**Comment by breeswish** [comment]
@andylokandy @iosmanthus PTAL
@mapleFU Please resolve conflicts.

**Comment by mapleFU** [comment]
I'll close this pr and create a new pr for streaming.
